<a href="https://colab.research.google.com/github/SaamiyaA123/VirtualWallet/blob/main/UB23_2022_Assignment1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Assignment 1 (SQL)

Answer the questions by modifying the lines that says 

```
REPLACE THIS LINE WITH YOUR ANSWER TO QUESTION XX
```

Each answer must be a single query. 

**DO NOT MODIFY ANY OTHER LINES  (INCLUDING %%read_sql dfXX LINE). CHANGING OTHER LINES INCREASES THE GRADER'S WORKLOAD AND RESULTS IN A REDUCED SCORE.**

For the detailed instructions, please refer "Assignment 1 Instructions" slides that is available from brightspace announcements.

In [2]:
# Installing related libraries to Colab.
!sudo apt-get install python3-dev libmysqlclient-dev > /dev/null
!pip install mysqlclient > /dev/null
!sudo pip3 install -U sql_magic > /dev/null

In [3]:
# for answer user
!wget https://www.dropbox.com/s/klf5a06c1poc7c2/ub23-client-cert.pem?dl=0 -O client-cert.pem -o /dev/null
!wget https://www.dropbox.com/s/y5b72ig82qmgt6r/ub23-client-key.pem?dl=0 -O client-key.pem -o /dev/null
!wget https://www.dropbox.com/s/ysrd8yduqx2xsgs/ub23-server-ca.pem?dl=0 -O server-ca.pem -o /dev/null

In [4]:
from sqlalchemy import create_engine

conn_string = 'mysql://{user}:{password}@{host}/{database}?charset=utf8'.format(
    host = '34.148.116.77', 
    user = 'ub23sqla',
    password = 'student',
    database = '',
    ssl={'ca':'./server-ca.pem', 'cert':'./client-cert.pem', 'key':'./client-key.pem'})

engine = create_engine(conn_string)
con = engine.connect()

In [5]:
# Prepare sql_magic library that enable to query to database easily.
%reload_ext sql_magic
%config SQL.conn_name = 'engine'

/usr/local/lib/python3.7/dist-packages/psycopg2/__init__.py:144: UserWarning: The psycopg2 wheel package will be renamed from release 2.8; in order to keep installing from binary please use "pip install psycopg2-binary" instead. For details see: <http://initd.org/psycopg/docs/install.html#binary-install-from-pypi>.
  """)


In [6]:
%%read_sql
use world;

Query started at 09:58:41 PM UTC; Query executed in 0.01 m

In [7]:
submission_mode = "s2"

### ERD of the world database

[ERD-World](http://4.bp.blogspot.com/-BUJLojNo4ik/UNIqWTKGwHI/AAAAAAAAAAY/DA4YD6EGJh4/s1600/world_erd.png)

### QUESTIONS Start here


Q1. List all the tables in the world database (10 pts)

In [7]:
%%read_sql df_q1
SHOW TABLES;

Query started at 09:01:43 PM UTC; Query executed in 0.00 m

Tables_in_world
0             city
1          country
2  countrylanguage

Sample answer


|index|Tables\_in_world|
|---|---|
|0|city|
|1|country|
|2|countrylanguage|



Q2. Describe the country table.

In [15]:
%%read_sql df_q2
DESCRIBE country;

Query started at 09:15:20 PM UTC; Query executed in 0.00 m

Field                                               Type Null  \
0             Code                                            char(3)   NO   
1             Name                                           char(52)   NO   
2        Continent  enum('Asia','Europe','North America','Africa',...   NO   
3           Region                                           char(26)   NO   
4      SurfaceArea                                      decimal(10,2)   NO   
5        IndepYear                                           smallint  YES   
6       Population                                                int   NO   
7   LifeExpectancy                                       decimal(3,1)  YES   
8              GNP                                      decimal(10,2)  YES   
9           GNPOld                                      decimal(10,2)  YES   
10       LocalName                                           char(45)   NO   
11  GovernmentForm                                           char(45)   NO   
12     HeadOfState                                           char(60)  YES   
13         Capital                                                int  YES   
14           Code2                                            char(2)   NO   

    Key Default Extra  
0   PRI                
1                      
2          Asia        
3                      
4          0.00        
5          None        
6             0        
7          None        
8          None        
9          None        
10                     
11                     
12         None        
13         None        
14

Sample answer

|index|Field|Type|Null|Key|Default|Extra|
|---|---|---|---|---|---|---|
|0|Code|char\(3)|NO|PRI|||
|1|Name|char\(52)|NO||||
|2|Continent|enum\('Asia','Europe','North America','Africa','Oceania','Antarctica','South America')|NO||Asia||
...
|13|Capital|int|YES||||
|14|Code2|char\(2)|NO||||

Q3. Obtain the number of rows in the country table

In [10]:
%%read_sql df_q3
SELECT COUNT(*) FROM country;

Query started at 09:08:07 PM UTC; Query executed in 0.00 m

COUNT(*)
0       239

Sample answer

|index|count\(*)|
|---|---|
|0|239|

Q4. List all names of the countries where 10 or more cities are registered

In [100]:
%%read_sql df_q4
SELECT country.Name FROM country 
JOIN city ON city.countrycode=country.code
GROUP BY city.countrycode 
HAVING COUNT(*) >= 10;

Query started at 12:24:46 AM UTC; Query executed in 0.00 m

Name
0            Argentina
1               Brazil
2               Canada
3                Chile
4                China
5             Colombia
6              Germany
7                Egypt
8                Spain
9       United Kingdom
10           Indonesia
11               India
12                Iran
13               Italy
14               Japan
15         South Korea
16              Mexico
17             Nigeria
18            Pakistan
19         Philippines
20              Poland
21  Russian Federation
22              Turkey
23              Taiwan
24             Ukraine
25       United States

Sample answer:

|index|Name|
|---|---|
|0|Argentina|
|1|Brazil|
|2|Canada|
...
|23|Taiwan|
|24|Ukraine|
|25|United States|

Q5. Obtain the smallest country for each continent in terms of SurfaceArea


In [101]:
%%read_sql df_q5
SELECT Name, Continent FROM country
WHERE SurfaceArea IN (SELECT MIN(SurfaceArea) FROM country GROUP BY continent);



Query started at 12:25:14 AM UTC; Query executed in 0.00 m

Name      Continent
0                         Bermuda  North America
1                   Bouvet Island     Antarctica
2                Falkland Islands  South America
3  British Indian Ocean Territory         Africa
4                           Macao           Asia
5                         Tokelau        Oceania
6   Holy See (Vatican City State)         Europe

Sample answer


|index|Name|Continent|
|---|---|---|
|0|Bermuda|North America|
|1|Bouvet Island|Antarctica|
...
|6|Holy See \(Vatican City State)|Europe|


Q6. Obtain the names of the country with the top-10 GNP boost (from top-1 to top-10). Here, GNP boost of a country is defined as GNP/GNPOld. Ignore the country where either GNP or GNPOld is Null.

In [102]:
%%read_sql df_q6
SELECT Name,GNP/GNPOld AS GNPboost FROM country
ORDER BY GNP/GNPOld DESC
LIMIT 10;

Query started at 12:25:29 AM UTC; Query executed in 0.00 m

Name  GNPboost
0  Congo, The Democratic Republic of the  2.814875
1                           Turkmenistan  2.198500
2                             Tajikistan  1.884470
3                                Estonia  1.580540
4                                Albania  1.282000
5                               Suriname  1.232295
6                                   Iran  1.222259
7                               Bulgaria  1.197561
8                               Honduras  1.135406
9                                 Latvia  1.134598

|index|Name|GNPboost|
|---|---|---|
|0|Congo, The Democratic Republic of the|2\.814875|
|1|Turkmenistan|2\.1985|
...
|8|Honduras|1\.135406|
|9|Latvia|1\.134598|

Q7: Obtain Name, Language and percentage in Africa that have 82% or more speakers of a single non-English language. Do **NOT** use subquery/CTE.


In [65]:
%%read_sql df_q7
SELECT country.Name,countrylanguage.Language, SUM(Percentage) AS percentage FROM country 
JOIN countrylanguage ON country.code = countrylanguage.CountryCode
WHERE countrylanguage.Language != 'English'
AND continent = 'Africa'
GROUP BY country.Name,countrylanguage.Language 
HAVING SUM(Percentage)>=82
ORDER BY SUM(Percentage) DESC
;


Query started at 11:25:47 PM UTC; Query executed in 0.00 m

Name Language  percentage
0      Cape Verde  Crioulo       100.0
1  Western Sahara   Arabic       100.0
2           Egypt   Arabic        98.8
3         Somalia   Somali        98.3
4         Burundi  Kirundi        98.1
5       Swaziland    Swazi        89.9

Sample Answer

|index|Name|Language|percentage|
|---|---|---|---|
|0|Cape Verde|Crioulo|100\.0|
|1|Western Sahara|Arabic|100\.0|
...
|4|Burundi|Kirundi|98\.1|
|5|Swaziland|Swazi|89\.9|